plugin/plugin.go

```diff
@@ -0,0 +1,369 @@
+		tiPlugins.versions[component] = version
+	}
+
+	// load plugin dl & manifest.
+	for _, pluginID := range cfg.Plugins {
+		var pName string
+		pName, _, err = ID(pluginID).Decode()
+		if err != nil {
+			return
+		}
+		// check duplicate.
+		_, dup := tiPlugins.versions[pName]
+		if dup {
+			if cfg.SkipWhenFail {
+				continue
+			}
+			err = errors.Errorf("plugin [%s] be duplicate declared in init parameters", pluginID)
+			return
+		}
+		// load dl.
+		var plugin Plugin
+		plugin, err = loadOne(cfg.PluginDir, ID(pluginID))
+		if err != nil {
+			if cfg.SkipWhenFail {
+				continue
+			}
+			return
+		}
+		tiPlugins.add(plugin)
+	}
+
+	// cross validate & load plugins.
+	for kind := range tiPlugins.plugins {
+		for i := range tiPlugins.plugins[kind] {
+			if err = tiPlugins.plugins[kind][i].validate(ctx, tiPlugins, initMode); err != nil {
+				if cfg.SkipWhenFail {
+					tiPlugins.plugins[kind][i].State = Disable
+					err = nil
+					continue
+				}
+				return
+			}
+			p := tiPlugins.plugins[kind][i]
+			if err = p.OnInit(ctx, p.Manifest); err != nil {
+				if cfg.SkipWhenFail {
+					tiPlugins.plugins[kind][i].State = Disable
+					err = nil
+					continue
+				}
+				return
+			}
+			if cfg.GlobalSysVar != nil {
+				for key, value := range tiPlugins.plugins[kind][i].SysVars {
+					(*cfg.GlobalSysVar)[key] = value
+					if value.Scope != variable.ScopeSession && cfg.PluginVarNames != nil {
+						*cfg.PluginVarNames = append(*cfg.PluginVarNames, key)
+					}
+				}
+			}
+			tiPlugins.plugins[kind][i].State = Ready
+		}
+	}
+	pluginGlobal = copyOnWriteContext{tiPlugins: unsafe.Pointer(tiPlugins)}
+	err = nil
+	return
+}
+
+func loadOne(dir string, pluginID ID) (plugin Plugin, err error) {
+	plugin.Path = filepath.Join(dir, string(pluginID)+LibrarySuffix)
+	plugin.library, err = gplugin.Open(plugin.Path)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	manifestSym, err := plugin.library.Lookup(ManifestSymbol)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	manifest, ok := manifestSym.(func() *Manifest)
+	if !ok {
+		err = errors.Errorf("can not read plugin %s's manifest", string(pluginID))
+		return
+	}
+	pName, pVersion, err := pluginID.Decode()
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	plugin.Manifest = manifest()
+	if plugin.Name != pName {
+		err = errors.Errorf("plugin load with %s but got wrong name %s", string(pluginID), plugin.Name)
+		return
+	}
+	if strconv.Itoa(int(plugin.Version)) != pVersion {
+		err = errors.Errorf("plugin load with %s but got %s", string(pluginID), strconv.Itoa(int(plugin.Version)))
+		return
+	}
+	return
+}
+
+// Reload hot swap a old plugin with new version.
+// Limit: loaded plugins shouldn't be unload and only be mark dying.
+func Reload(ctx context.Context, cfg Config, pluginID ID) (err error) {
+	newPlugin, err := loadOne(cfg.PluginDir, pluginID)
+	if err != nil {
+		return
+	}
+	_, err = replace(ctx, cfg, newPlugin.Name, newPlugin)
```

Comment:
Reload may execute by multiple session simultaneously?
Then only one of them will success because the atomic copy on write?